kernel/src/hil/flash.rs

```diff
@@ -0,0 +1,24 @@
+pub enum Error {
+}
+
+pub type FlashResult = Result<(), Error>;
+
+/// A block of readable persistent flash memory.
+pub trait FlashReadable {
+    type Word;
+
+    /// Gets the size of the flash memory in words.
+    fn len(&self) -> usize;
+
+    /// Reads into a buffer from flash memory.
+    fn read(&self, offset: usize, buf: &mut [Self::Word]) -> FlashResult;
+}
+
+/// A block of writable persistent flash memory.
+pub trait FlashWritable: FlashReadable {
+    /// Writes the contents of a buffer to flash memory.
+    fn write(&self, offset: usize, buf: &[Self::Word]) -> FlashResult;
```

Comment:
That's a good point, it's not well-documented yet (or at all). Again, there are two parts, the asynchronous interface _in_ the kernel (which I think is what matters for this) and the asynchronous interface with processes (they have different isolation models, so the asynchronous "communication" is also different). I'll write something up in the next few days and send it to the mailing list. But for now, a good example is the UART hil and corresponding implementations (e.g. `chips/sam4l/src/usart.rs`) and the Alarm hil.

The basic idea is you have one trait for the controller (e.g. the FlashWriteable hil) and another trait for the client which has callback methods (e.g. `write_done`). Then, you configure each to reference each other (that's basically why we need all the `self` references to be shared).